src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage_FlushWrites.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SQLite.v2.Interop;
+
+namespace Microsoft.CodeAnalysis.SQLite.v2
+{
+    internal partial class SQLitePersistentStorage
+    {
+        private readonly object _flushGate = new object();
+
+        /// <summary>
+        /// Task kicked off to actually do the work of flushing all data to the DB. If we hear about
+        /// new writes to the storage system we don't have to kick off another flush task if one is
+        /// active.
+        /// </summary>
+        private Task _flushTask;
+
+        private void EnqueueFlushTask()
+        {
+            lock (_flushGate)
+            {
+                // Check if we already have a flush task in flight.  If so, no need to make another.
+                if (_flushTask == null)
+                {
+                    // We're performing writes, so we need to enqueue this work to the exclusive
+                    // scheduler.  Because this happens in the future, we may end up getting run
+                    // after the storage system has shutdown.  So we need to check to make sure we
+                    // should still continue once we wake up.
+                    var token = _shutdownTokenSource.Token;
+                    _flushTask = Task.Delay(FlushAllDelayMS, token).ContinueWith(
+                        s_flushInMemoryDataToDiskIfNotShutdown, this, token, TaskContinuationOptions.None, _readerWriterLock.ExclusiveScheduler);
+                }
+            }
+        }
+
+        private static readonly Action<Task, object> s_flushInMemoryDataToDiskIfNotShutdown = (Task _1, object self) =>
+        {
+            var storage = (SQLitePersistentStorage)self;
+
+            lock (storage._flushGate)
+            {
+                // Indicate that there is no outstanding write task.  The next request to 
+                // write will cause one to be kicked off.
+                storage._flushTask = null;
+
+                if (storage._shutdownTokenSource.IsCancellationRequested)
+                {
+                    // Don't flush from a bg task if we've been asked to shutdown.  The shutdown
+                    // logic in the storage service will take care of the final writes to the main
+                    // db.
+                    return;
+                }
+
+                // Haven't been shutdown.  Actually go and move any outstanding data to the real DB.
+                using var _ = storage.GetPooledConnection(out var connection);
+                storage.FlushInMemoryDataToDisk_MustRunUnderLock(connection);
```

Comment:
We're doing the actual flushing under _flushGate? I would have imagined this was kept outside as otherwise any write to the in-memory is still going to block behind this writing. Presumably the SQLite transaction semantics already make it safe (otherwise this has bigger problems), and since this is running after the nulling-out of the task it means if this runs outside the gate there's a possibility of an extra flush job being queued, but that won't do anything so it's moot.

Reply:
great find.  I think this is actually a pure bug (though not likely to hit).  I think we likely should not have this flush gate tbh.  we already have a mechanism to ensure exclusive access through the threading mechanism.